packages/cli/src/executions/execution-recovery.service.ts

```diff
@@ -0,0 +1,189 @@
+import Container, { Service } from 'typedi';
+import { Push } from '@/push';
+import { sleep } from 'n8n-workflow';
+import { ExecutionRepository } from '@db/repositories/execution.repository';
+import { getWorkflowHooksMain } from '@/WorkflowExecuteAdditionalData'; // @TODO: Dependency cycle
+import { InternalHooks } from '@/InternalHooks'; // @TODO: Dependency cycle if injected
+import type { DateTime } from 'luxon';
+import type { IRun, ITaskData } from 'n8n-workflow';
+import type { EventMessageTypes } from '../eventbus/EventMessageClasses';
+import type { IExecutionResponse } from '@/Interfaces';
+import { NodeCrashedError } from '@/errors/node-crashed.error';
+import { WorkflowCrashedError } from '@/errors/workflow-crashed.error';
+import { ARTIFICIAL_TASK_DATA } from '@/constants';
+
+/**
+ * Service for recovering executions truncated by an instance crash.
+ */
+@Service()
+export class ExecutionRecoveryService {
+	constructor(
+		private readonly push: Push,
+		private readonly executionRepository: ExecutionRepository,
+	) {}
+
+	/**
+	 * "Recovery" means (1) amending key properties of a truncated execution,
+	 * (2) running post-execution hooks, and (3) returning the amended execution
+	 * so the UI can reflect the error. "Recovery" does **not** mean injecting
+	 * execution data from the logs (they hold none), or resuming the execution
+	 * from the point of truncation, or re-running the whole execution.
+	 *
+	 * Recovery is only possible if event logs are available in the container.
+	 * In regular mode, logs should but might not be available, e.g. due to container
+	 * being recycled, max log size causing rotation, etc. In queue mode, as workers
+	 * log to their own filesystems, only manual exections can be recovered.
+	 */
+	async recover(executionId: string, messages: EventMessageTypes[]) {
+		if (messages.length === 0) return null;
+
+		const amendedExecution = await this.amend(executionId, messages);
+
+		if (!amendedExecution) return null;
+
+		await this.executionRepository.updateExistingExecution(executionId, amendedExecution);
+
+		await this.runHooks(amendedExecution);
+
+		this.push.once('editorUiConnected', async () => {
+			await sleep(1000);
+			this.push.broadcast('executionRecovered', { executionId });
+		});
+
+		return amendedExecution;
+	}
+
+	/**
+	 * Amend `status`, `stoppedAt`, and `data` of an execution using event log messages.
+	 */
+	private async amend(executionId: string, messages: EventMessageTypes[]) {
+		const { nodeMessages, workflowMessages } = this.toRelevantMessages(messages);
+
+		if (nodeMessages.length === 0) return null;
```

Comment:
In this case we're exiting the process prematurely and the execution would never be updated.

If `nodeMessages` were not found, we should just update to mark it as crashed immediately and then return, shouldn't we?

Reply:
Same reply as here: https://github.com/n8n-io/n8n/pull/9613#discussion_r1630716688